### PR TITLE
Notif overhaul

### PR DESCRIPTION
**Pull request by NEUDitao**
Hey y'all, been working on a notifications system overhaul a la https://github.com/ryanhugh/searchneu/issues/10. It's pretty functional (aside from a few memory leaks), but it's also really ugly. Dropping this here so people can start commenting on it!

### CODE REVIEW

File: common/important.rkt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#lang racket
```
Comment (ryanhugh) [inline common/important.rkt]: LOL

can you move this into the scripts folder?

and maybe we could have an racket-themed easter egg on the site - maybe like something comes up when you type in "racket"?  🥚 🤔

File: frontend/components/SignUpForNotifications.js [hunk 2]
```diff
@@ -36,6 +36,7 @@ class SignUpForNotifications extends React.Component {
     this.facebookScopeRef = null;
     this.onSubscribeToggleChange = this.onSubscribeToggleChange.bind(this);
     this.closeModal = this.closeModal.bind(this);
+    this.handleClick = this.props.handleClick;
```
Comment (ryanhugh) [inline frontend/components/SignUpForNotifications.js]: Lets not do this here - props can change when the component is updated (componentDidUpdate) - so saving this in the constructor would ignore any future updates of this prop. Lets just do this.props.handleClick throughout the component!

File: frontend/components/facebook.js [hunk 3]
```diff
@@ -93,12 +100,16 @@ class Facebook {
     } else if (e.event === 'opt_in') {
       macros.log('Opt in was clicked!', e);
 
+      user.downloadUserData(100);
```
Comment (ryanhugh) [inline frontend/components/facebook.js]: Maybe we can have some better logic than retry 100 times? I'd have to think about this though.. is a bit yucky

File: frontend/components/facebook.js [hunk 4]
```diff
@@ -120,7 +131,7 @@ class Facebook {
                     timestamp: Date.now(),
                     sender:
                       {
-                        id: '1397905100304615',
+                        id: '2178896222126069',
```
Comment (ryanhugh) [inline frontend/components/facebook.js]: Hmm - I wonder if there would be any way to make it so it could pull this from config.json? or something that supports multiple IDs

File: frontend/components/panels/DesktopClassPanel.js [hunk 5]
```diff
@@ -24,16 +25,32 @@ import globe from './globe.svg';
...
     macros.debounceTooltipRebuild();
+    user.registerCallback(this.reRenderCheckBox);
+  }
+
+  componentWillUnmount() {
+    user.unregisterCallback(this.reRenderCheckBox);
+  }
+
+  reRenderCheckBox() {
+    this.setState({ test: !this.state.test });
```
Comment (ryanhugh) [inline frontend/components/panels/DesktopClassPanel.js]: test ?

File: frontend/components/panels/DesktopClassPanel.js [hunk 6]
```diff
@@ -73,6 +90,18 @@ export default class DesktopClassPanel extends BaseClassPanel {
     return false;
   }
 
+
+  // handles the sign up for notifications button being clicked,
+  // to allow for user to choose which sections to sign up for
+  handleClick() {
+    // fishy...
+    this.state.showNotificationBoxes = true;
```
Comment (ryanhugh) [inline frontend/components/panels/DesktopClassPanel.js]: lets clean this up :) lemme know if ya want me to take a look! :)

File: backend/server.js [hunk 7]
```diff
@@ -581,10 +546,12 @@ app.post('/webhook/', wrap(async (req, res) => {
 
       if (text === 'test') {
         notifyer.sendFBNotification(sender, 'CS 1800 now has 1 seat available!! Check it out on https://searchneu.com/cs1800 !');
-      } else if (text.toLowerCase() === 'stop') {
-        unsubscribeSender(sender);
+      } else if (text.toLowerCase().includes('stop')) {
+        unsubscribeSender(sender, text);
       } else if (text === 'What is my facebook messenger sender id?') {
         notifyer.sendFBNotification(sender, sender);
+      } else if (text === 'no u') {
```
[the remark this answers is not in the document]
Comment (NEUDitao) [reply backend/server.js]: no u

File: backend/server.js [hunk 8]
```diff
@@ -581,10 +546,12 @@ app.post('/webhook/', wrap(async (req, res) => {
 
       if (text === 'test') {
         notifyer.sendFBNotification(sender, 'CS 1800 now has 1 seat available!! Check it out on https://searchneu.com/cs1800 !');
-      } else if (text.toLowerCase() === 'stop') {
-        unsubscribeSender(sender);
+      } else if (text.toLowerCase().includes('stop')) {
```
[the remark this answers is not in the document]
Comment (NEUDitao) [reply backend/server.js]: it was gonna be a hack going back

File: common/important.rkt (hunk 1 above)
Comment (NEUDitao) [reply common/important.rkt]: :o

File: frontend/components/facebook.js (hunk 3 above)
Comment (NEUDitao) [reply frontend/components/facebook.js]: 🤔 Will sit on this for now, and try to look at the othe rother thing

File: frontend/components/panels/DesktopClassPanel.js (hunk 5 above)
Comment (NEUDitao) [reply frontend/components/panels/DesktopClassPanel.js]: it was to re render the checkbox... is there a more concrete way than this?

File: frontend/components/panels/DesktopClassPanel.js (hunk 6 above)
Comment (NEUDitao) [reply frontend/components/panels/DesktopClassPanel.js]: will try to figure out... don't know why it was angry last time, might hcnage now?

File: frontend/components/panels/NotifCheckBox.js [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+    this.state = {};
+    this.state.checked = user.hasSectionAlready(Keys.getSectionHash(wantedSection));
+    this.state.section = wantedSection;
+  }
+
+  // if the state is currently checked, uncheck, remove the section from the user's data
+  // do opposite.
+  // send data to backend
+  handleChange = () => {
```
Comment (ryanhugh) [inline frontend/components/panels/NotifCheckBox.js]: I know this is a p clever method syntax that auto-binds `this` to the method, but can you do the normal syntax and then bind in the constructor? Last time i checked sublime doesn't index this as a method and we should keep code consistency.... :/

File: frontend/components/panels/NotifCheckBox.js [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    if (this.state.checked) {
+      user.removeSection(this.state.section);
+      this.setState({ checked: false });
+    } else {
+      user.enrollSection(this.state.section);
+      this.setState({ checked: true });
+    }
+    user.sendUserData();
```
Comment (ryanhugh) [inline frontend/components/panels/NotifCheckBox.js]: Awesome!! 😍

File: frontend/components/panels/NotifCheckBox.js [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+import React from 'react';
+import { Checkbox } from 'semantic-ui-react';
+import user from '../user';
+import macros from '../macros';
+import Keys from '../../../common/Keys';
+
+// This file renders the checkboxes that control which sections a user signs up for
+// notifications.
+
+export default class NotifCheckBox extends Checkbox {
```
Comment (ryanhugh) [inline frontend/components/panels/NotifCheckBox.js]: Extend React.component - like all the other components - and then just use checkbox in the render method

File: frontend/components/panels/NotifCheckBox.js (hunk 10 above)
Comment (NEUDitao) [reply frontend/components/panels/NotifCheckBox.js]: :o

File: frontend/components/panels/NotifCheckBox.js (hunk 9 above)
Comment (NEUDitao) [reply frontend/components/panels/NotifCheckBox.js]: just emaCz

**Comment by NEUDitao** [comment]
Tests are done. So the big issues still around

  - The await get(100) thing is just super sus. Tackling the slow download could be a thing, but that might be a while. We'll see though
  - That one pesky memory leak that's still around... somewhere. Haven't found it in a bit, but still
  - Mobile

**Comment by NEUDitao** [comment]
Refactored the work flow as requested through Dms, again, cannot test on mobile because cOmPuTeR is dumbo

**Comment by NEUDitao** [comment]
As discussed w/ Ryan, having a weird issue w/ FireFox on my end. Can anyone else test is signing up for notifications is currently broken on FF? (window.FB is not loading).

Also cannot test mobile properly on Chrome, though in theory it works. Working on that on my end currently.

**Comment by edward-shen** [comment]
Firefox has stronger encapsulation and stronger privacy defaults, including blocking trackers. It's possible that it's seeing FB on a non-FB site and blocking that.

Users might also face issues when they install the Mozilla-recommended Facebook Container, which encapsulates Facebook from the browser.

**Comment by ryanhugh** [comment]
The live site works in my firefox. Lets look into this a bit more before throwing up a warning for ff - I think we can get it working :)  I'll PM ya!

image

image

**Comment by ryanhugh** [comment]
Looks like the site doesn't work if FF is in strict Content Blocking mode - perhaps in a future PR we could throw up a warning if this is enabled

image

**Comment by ryanhugh** [comment]
Status of stuff:

 - I pushed a fix for the firefox thing in a different PR #85 
 - I also made a bunch of changes to the sender id stuff #85 (lets keep the functionality in #85)
 - Removed `/getUserData` API and a bunch of code in user.js, which we will need for this PR and we can add it back again.

Todo here:

- Cleanup the callback stuff
- and the retry 100 stuff (will need to add back and change the `/getUserData` code i just removed)
- some other cleanup
- mobile support

Lmk if you wanna do this and I'll give some pointers!

File: backend/server.js [hunk 12]
```diff
@@ -693,6 +704,60 @@ async function findMatchingUser(requestLoginKey) {
...
+  if (!req.body || !req.body.loginKey) {
+    res.send(JSON.stringify({
+      error: 'Error.',
+    }));
+    return;
+  }
+
+  // Checks checks checks
+  // Make sure the login key is valid
+  if (typeof req.body.loginKey !== 'string' || req.body.loginKey.length !== 100) {
```
Comment (ryanhugh) [inline backend/server.js]: We need to make sure

Comment (ryanhugh) [inline backend/server.js]: We need to make sure that the given user loginKey actually matches one of the keys that are in the DB, too ;)

File: frontend/components/SignUpForNotifications.js [hunk 13]
```diff
@@ -47,7 +59,16 @@ class SignUpForNotifications extends React.Component {
       return;
     }
 
-    window.FB.XFBML.parse(this.facebookScopeRef);
+    // firefox has a strict no-trackers rule, which FB violates
+    // this is to make sure the site doesn't crash on firefox
+    if (window.FB) {
+      window.FB.XFBML.parse(this.facebookScopeRef);
+    } else {
+      this.setState({ // eslint-disable-line react/no-did-update-set-state
```
Comment (ryanhugh) [inline frontend/components/SignUpForNotifications.js]: Lets not use `react/no-did-update-set-state` - lets figure out the right way to do this!

File: frontend/components/facebook.js (hunk 4 above)
Comment (ryanhugh) [reply frontend/components/facebook.js]: I actually fixed this in a different PR too i think - sync the latest master!

File: backend/search.js [hunk 14]
```diff
@@ -385,6 +385,7 @@ class Search {
       // https://github.com/ryanhugh/searchneu/pull/19
       numerica: 'numerical',
       numeri: 'numerical',
+      swd: 'software dev',
```
Comment (ryanhugh) [inline backend/search.js]: DJ is working on moving over core search to elasticsearch, drop a comment about `swd` here! https://github.com/ryanhugh/searchneu/pull/87/files#diff-d74b9dd40c18a857e1ddbac80c22003cR90

File: frontend/components/Home.js [hunk 15]
```diff
@@ -114,6 +115,7 @@ class Home extends React.Component {
 
       // Keep track of whether the help modal is open or not.
       helpModalOpen: false,
+
```
Comment (ryanhugh) [inline frontend/components/Home.js]: can you remove these added newlines?

File: frontend/components/SignUpForNotifications.js [hunk 16]
```diff
@@ -17,23 +17,32 @@ import user from './user';
...
-      // Keeps track of whether the adblock message is being shown or not
-      // Sometimes, adblock will block the FB plugin from loading
-      // Firefox strict browsing also blocks this plugin from working
-      // If the plugin failed to load for whatever reason, show this message and ask the user to allow FB plugins
+      // or the adblock message (sometimes, adblock blocks fb messenger button)
       showAdblockMessage: false,
+
+      // if firefox has strict privacy tracking enabled, the page can crash. Instead of
+      // doing that, we make sure we dont' try to render null things.
+      showFirefoxMessage: false,
```
Comment (ryanhugh) [inline frontend/components/SignUpForNotifications.js]: and remove this firefox stuff - this was fixed in another PR

File: frontend/components/SignUpForNotifications.js [hunk 17]
```diff
@@ -215,9 +229,10 @@ class SignUpForNotifications extends React.Component {
       <div className='sign-up-for-notifications-container'>
         {content}
         <Modal
+          className='adblock-notification-modal-container'
           header='Please disable adblock and sign into Facebook.'
           open={ this.state.showAdblockMessage }
-          content="Please disable any ad blocking extentions for this site because this feature does not work when adblock is enabled. If you are using Firefox in strict blocking mode, you will need to add an exception for this site for this feature to work. You can also try using a different browser. If you can't get it working send me a message at ryanhughes624@gmail.com."
+          content="Please open a new tab and ensure you are signed into Facebook. Also, make sure to disable any ad blocking extentions because this feature does not work when adblock is enabled. If you are using Safari, try using a different browser. If you can't get it working send me a message at ryanhughes624@gmail.com."
```
Comment (ryanhugh) [inline frontend/components/SignUpForNotifications.js]: revert this change

**Comment by NEUDitao** [comment]
Some updates, since I haven't had a while.

1. This feature's taken a while! That's because from May to August I didn't have a computer, or internet consistently! Picking up work where it dropped off was difficult.
2. The feature works now. If you want to test it (please test it), just checkout the branch, make sure elasticsearch is running, and run it locally. 
3. no u.
4. There still isn't a version of long polling done yet. That could probably be done in the future, but getting the feature out might be more valuable.

**Comment by NEUDitao** [comment]
Pulled a funny prank. Implemented long polling.

**Comment by ryanhugh** [comment]
Fix up a bunch of stuff on travis when you have a second! `yarn fix`

File: backend/server.js [hunk 18]
```diff
@@ -305,41 +311,30 @@ async function onSendToMessengerButtonClick(sender, userPageId, b64ref) {
...
+      }
+      existingData.watchingSections = _.uniq(existingData.watchingSections.concat(userObject.sectionHashes));
     }
-
-    existingData.watchingSections = _.uniq(existingData.watchingSections.concat(userObject.sectionHashes));
-
+*/
     // Remove any null or undefined values from the watchingClasses and watchingSections
-    // This can happen if data is manually deleted from the DB, and the data is no longer contineous.
+    // This can happen if data is manually deleted from the DB, and the data is no longer continuous.
```
Comment (ryanhugh) [inline backend/server.js]: haha thanks

File: backend/server.js [hunk 19]
```diff
@@ -305,41 +311,30 @@ async function onSendToMessengerButtonClick(sender, userPageId, b64ref) {
...
-    } else if (sectionWasentWatchingBefore.length === 0) {
-      notifyer.sendFBNotification(sender, `Successfully signed up for notifications if sections are added to ${classCode}!`);
-    } else {
-      notifyer.sendFBNotification(sender, `Successfully signed up for notifications for ${sectionWasentWatchingBefore.length} sections in ${classCode}!`);
-    }
 
-    // ok lets add what classes the user saw in the frontend that have no seats availible and that he wants to sign up for
-    // so pretty much the same as courspro - the class hash and the section hashes - but just for the sections that the user sees that are empty
-    // so if a new section is added then a notification will be send off that it was added but the user will not be signed up for it
+    // only auto enrolls if there's one (or less???) sectionz remaining
```
Comment (ryanhugh) [inline backend/server.js]: yup - one or zero sections. If a class has no sections and the user clicks sign up, they can be notified if sections are added, or some class details change

File: frontend/components/Home.js (hunk 15 above)
Comment (ryanhugh) [reply frontend/components/Home.js]: this

File: frontend/components/SignUpForNotifications.js (hunk 16 above)
Comment (ryanhugh) [reply frontend/components/SignUpForNotifications.js]: this

File: frontend/components/SignUpForNotifications.js (hunk 13 above)
Comment (ryanhugh) [reply frontend/components/SignUpForNotifications.js]: ping

**Comment by ryanhugh** [comment]
Great work Eddy!!

@dajinchu Take a look when ya have a chance 

Here's some comments

Pasted Graphic


Can you expand this message to clarify what happened and tell the user what they are signed up for? Like, when I first click the send to messenger button am I signed up for notifications on the class or not - eg will I be notified if sections are added and removed from the class? 

https://github.com/ryanhugh/searchneu/blob/master/backend/updater.js#L218


Also should we send a FB notification when they click the switch to tell them they are signed up for a new section? 



Pasted Graphic 1


Can you make the (i) buttons smaller/a lighter color to de-emphasize it compared to the other ui stuff?


When you first refresh the page, should we make it show the switches for classes people are signed up for? (instead of saying click here to sign up for notifications when they already did that and we know they did)


Pasted Graphic 2

When there isn’t any sections, lets not say toggle the sections you want to be signed up for.

Make sure to go over all the edge cases and think about the user experience when using the feature in different scenarios - like what information are we telling the user when they take certain actions? :) 

good work!

**Comment by dajinchu** [comment]
I agree with Ryan on the usability. Looks good on the elasticsearch end of things!

File: frontend/components/panels/NotifCheckBox.js [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+  // only, otherwise, set up callback on onChange
+  render() {
+    // one last check to ensure the state is correct... since sometimes the user isn't
+    // in place by the time rendering has started.
+    if (this.state.checked !== user.hasSectionAlready(Keys.getSectionHash(this.state.section))) {
+      this.state.checked = !this.state.checked;
+    }
+
+    // if we have a section, and the section has seats remaining, it doesn't make
+    // sense to sig up a user for notifications, so make the Checkbox readonly
```
Comment (ryanhugh) [inline frontend/components/panels/NotifCheckBox.js]: outdated comment

File: frontend/components/panels/DesktopSectionPanel.js [hunk 21]
```diff
@@ -116,6 +135,22 @@ export default class DesktopSectionPanel extends React.Component {
 
     const honorsCheck = this.props.section.honors ? <Icon name='check' /> : <Icon name='x' />;
 
+    let notifBox;
+
+    if (this.props.section.seatsRemaining === 0) {
+      notifBox = (
+        <div data-tip='Sign up for notifications for this section' className='inlineBlock'>
```
Comment (ryanhugh) [inline frontend/components/panels/DesktopSectionPanel.js]: Can you move this <div data-tip='... > into the < NotifCheckBox > file? Would make more sense to have this logic inside this file, especially if you passing this.props.section.seatsRemaining into < NotifCheckBox >

File: frontend/components/panels/DesktopSectionPanel.js [hunk 22]
```diff
@@ -116,6 +135,22 @@ export default class DesktopSectionPanel extends React.Component {
...
+          <NotifCheckBox seats={ this.props.section.seatsRemaining } section={ this.props.section } />
```
Comment (ryanhugh) [inline frontend/components/panels/DesktopSectionPanel.js]: no need to pass both this.props.section.seatsRemaining and this.props.section, because you can get the seatsRemaining from this.props.section

File: frontend/components/panels/NotifCheckBox.js [hunk 23]
```diff
@@ -0,0 +1,76 @@
...
+  // is currently in the user or not
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      // has a user already signed up for notifications on a section?
+      checked: user.hasSectionAlready(Keys.getSectionHash(this.props.section)),
+
+      // the section for a class that this NotifCheckBox is for
+      section: this.props.section,
```
Comment (ryanhugh) [inline frontend/components/panels/NotifCheckBox.js]: Don't copy section to the state - throughout this component you want to just use this.props.section. 

Give these docs a read-down, they've got a lot of info about how react works :) 

https://reactjs.org/docs/components-and-props.html

File: frontend/components/panels/NotifCheckBox.js [hunk 24]
```diff
@@ -0,0 +1,76 @@
...
+    user.sendUserData();
+  }
+
+
+  // renders the proper checkbox. If there are still seats, then make it read
+  // only, otherwise, set up callback on onChange
+  render() {
+    // one last check to ensure the state is correct... since sometimes the user isn't
+    // in place by the time rendering has started.
+    if (this.state.checked !== user.hasSectionAlready(Keys.getSectionHash(this.state.section))) {
```
Comment (ryanhugh) [inline frontend/components/panels/NotifCheckBox.js]: I think you can remove this entire if statement - and just use 
```
user.hasSectionAlready(Keys.getSectionHash(this.state.section))
```
as the source of truth. 

As a side note, you don't want to modify this.state in render, and when you do modify this.state. you'd want to use `this.setState` and don't directly assign to `this.state`

Comment (NEUDitao) [reply frontend/components/panels/NotifCheckBox.js]: oh yeah that was a hack around for the beginning... will remove

File: backend/server.js [hunk 25]
```diff
@@ -538,6 +527,175 @@ app.post('/subscribeEmail', wrap(async (req, res) => {
...
+    const user = await database.get(`/users/${senderId}`);
+    if (user && !user.loginKeys.includes(req.body.loginKey)) {
+      macros.log('loginKey not within the sender\'s entry');
+      res.send(JSON.stringify({
+        error: 'Error, invalid loginkey on senderId.',
+      }));
+      return;
+    }
+    await database.set(`/users/${senderId}`, req.body.info);
+    macros.log('sending done, result is ', await database.get(`/users/${senderId}`));
```
Comment (ryanhugh) [inline backend/server.js]: remove this line when you're done with everything - the logging is find just don't want the DB call

File: frontend/components/SignUpForNotifications.js [hunk 26]
```diff
@@ -17,16 +17,24 @@ import user from './user';
 // (This Sent To Messenger button is not rendered initially because it requires that an iframe is added and 10+ http requests are made for each time it is rendered)
 
 // TODO: Lets make it so clicking on the Send To Messenger button changes this to a third state that just says thanks for signing up!
+// (still needs to be done!)
+
+// All of the actually signing up for sections is now in NotifCheckbox
 
 class SignUpForNotifications extends React.Component {
+  static hasAdblock = false;
```
Comment (ryanhugh) [inline frontend/components/SignUpForNotifications.js]: remove

File: backend/server.js [hunk 27]
```diff
@@ -365,6 +338,14 @@ async function onSendToMessengerButtonClick(sender, userPageId, b64ref) {
     loginKeys.add(userObject.loginKey);
     existingData.loginKeys = Array.from(loginKeys);
 
+    macros.log('neal young', reqs, userObject.loginKey, reqs[userObject.loginKey]);
+    reqs[userObject.loginKey].send((JSON.stringify({
```
Comment (ryanhugh) [inline backend/server.js]: What if `userObject.loginKey` doesn't exist on reqs? Want to surround this with a quick check like: 

```js
if (reqs[userObject.loginKey]) {...}
```

File: backend/server.js [hunk 28]
```diff
@@ -387,11 +368,19 @@ async function onSendToMessengerButtonClick(sender, userPageId, b64ref) {
...
-
     // Send the user a notification letting them know everything was successful.
-    notifyer.sendFBNotification(sender, `Thanks for signing up for notifications ${names.first_name}! I'll send you another message if a seat opens up in ${aClass.subject} ${aClass.classId}!`);
+    notifyer.sendFBNotification(sender, `Thanks for signing up for notifications ${names.first_name}! 
+
+Toggle the sliders back on https://searchneu.com/ to sign up for notifications!`);
 
     database.set(`/users/${sender}`, newUser);
+    macros.log('neal young', reqs, userObject.loginKey, reqs[userObject.loginKey]);
+    reqs[userObject.loginKey].send((JSON.stringify({
```
Comment (ryanhugh) [inline backend/server.js]: same here

File: backend/server.js [hunk 29]
```diff
@@ -538,6 +529,253 @@ app.post('/subscribeEmail', wrap(async (req, res) => {
...
+  const userObject = await sendUserHelper(req, res);
+
+  userObject.watchingSections.push(req.body.info);
+
+  await database.set(`/users/${req.body.senderId}`, userObject);
+  macros.log('sending done.');
+
+  const sectionInfo = req.body.sectionInfo;
+
+  notifyer.sendFBNotification(req.body.senderId, `Thanks for signing up for a section in ${sectionInfo.subject} ${sectionInfo.classId} (CRN: ${sectionInfo.crn}). Check it out at https://searchneu.com/${sectionInfo.termId}/${sectionInfo.subject}${sectionInfo.classId} !`);
```
Comment (ryanhugh) [inline backend/server.js]: We don't need to link back to the site when we send this msg - can you remove this part? 

```suggestion
  notifyer.sendFBNotification(req.body.senderId, `Thanks for signing up for a section in ${sectionInfo.subject} ${sectionInfo.classId} (CRN: ${sectionInfo.crn}).`);
```

File: backend/server.js [hunk 30]
```diff
@@ -538,6 +529,253 @@ app.post('/subscribeEmail', wrap(async (req, res) => {
...
+  // send a status of success. Hopefully it went well.
+  res.send(JSON.stringify({
+    status: 'Success',
+  }));
+}));
+
+app.post('/removeSection', wrap(async (req, res) => {
+  const userObject = await sendUserHelper(req, res);
+
+  macros.log(userObject);
```
Comment (ryanhugh) [inline backend/server.js]: remove this logging

File: backend/server.js [hunk 31]
```diff
@@ -538,6 +529,253 @@ app.post('/subscribeEmail', wrap(async (req, res) => {
...
+    status: 'Success',
+  }));
+}));
+
+app.post('/addClass', wrap(async (req, res) => {
+  const userObject = await sendUserHelper(req, res);
+  userObject.watchingClasses.push(req.body.info);
+
+  await database.set(`/users/${req.body.senderId}`, userObject);
+  macros.log('sending done.');
```
Comment (ryanhugh) [inline backend/server.js]: logging

File: backend/server.js [hunk 32]
```diff
@@ -538,6 +529,253 @@ app.post('/subscribeEmail', wrap(async (req, res) => {
...
+      }
+    }
+  }
+
+  return null;
+}
+
+
+// sends data to the database in the backend
+async function sendUserHelper(req, res) {
```
Comment (ryanhugh) [inline backend/server.js]: Want rename this to something like `verifyRequestAndGetDbUser` ? 

I know this is totally contradicting stuff taught at neu but I feel like naming functions "helper" isn't very informative to what they actually do.

File: backend/server.js [hunk 33]
```diff
@@ -538,6 +529,253 @@ app.post('/subscribeEmail', wrap(async (req, res) => {
...
+}
+
+
+app.post('/addSection', wrap(async (req, res) => {
+  const userObject = await sendUserHelper(req, res);
+
+  userObject.watchingSections.push(req.body.info);
+
+  await database.set(`/users/${req.body.senderId}`, userObject);
+  macros.log('sending done.');
```
Comment (ryanhugh) [inline backend/server.js]: same here

File: backend/server.js [hunk 34]
```diff
@@ -538,6 +529,253 @@ app.post('/subscribeEmail', wrap(async (req, res) => {
...
+  }
+
+  if (!acc) {
+    _.pull(userObject.watchingClasses, req.body.classHash);
+  }
+
+  const sectionInfo = req.body.sectionInfo;
+
+
+  macros.log(userObject);
```
Comment (ryanhugh) [inline backend/server.js]: logging

File: frontend/components/SignUpForNotifications.js [hunk 35]
```diff
@@ -215,6 +254,7 @@ class SignUpForNotifications extends React.Component {
       <div className='sign-up-for-notifications-container'>
         {content}
         <Modal
+          className='adblock-notification-modal-container'
```
Comment (ryanhugh) [inline frontend/components/SignUpForNotifications.js]: What's the purpose of  this ?

File: backend/server.js [hunk 36]
```diff
@@ -538,6 +529,253 @@ app.post('/subscribeEmail', wrap(async (req, res) => {
...
+  if (senderId) {
+    const user = await database.get(`/users/${senderId}`);
+
+    if (!user) {
+      if (reqs[req.body.loginKey]) {
+        reqs[req.body.loginKey].send(JSON.stringify({
+          error: 'Error, multiple requests from same user in quick succession',
+        }));
+      }
+      reqs[req.body.loginKey] = [res, new Date()];
```
Comment (ryanhugh) [inline backend/server.js]: So here, to access res, you would have to go `reqs[req.body.loginKey][0]`, but everywhere else you access res you just go `reqs[req.body.loginKey]`, like above `reqs[req.body.loginKey].send`. 

Want to fix this up here and elsewhere? and want to make it an object {res: res, receiveTime: Date()} instead of an array?

File: backend/server.js [hunk 37]
```diff
@@ -38,6 +38,12 @@ const app = express();
 // This is not used in development
 const fbAppSecret = macros.getEnvVariable('fbAppSecret');
 
+
+// All of the requests/responses that haven't been pushed yet. Added here when
+// requests come in for data that isn't quite in the backend yet
+// saved as a loginKey: res pair
```
Comment (ryanhugh) [inline backend/server.js]: Update this too, because you store timestamps in here too

**Comment by NEUDitao** [comment]
God's in his heaven, and all is right in the world.
